tests/integration/conftest.py

```diff
@@ -590,56 +590,59 @@ async def _complete(
     """
     start_time = time()
 
-    _tokens_list: List[Tokens] = []
-    for tokens in tokens_list:
+    tokens_list: List[Tokens] = []
+    for tokens in wait_tokens:
         if isinstance(tokens, str):
             tokens = Tokens(tokens, relative=True)
-        _tokens_list.append(tokens.task)
+        tokens_list.append(tokens.task)
 
     # capture task completion
     remove_if_complete = schd.pool.remove_if_complete
 
     def _remove_if_complete(itask, output=None):
-        nonlocal _tokens_list
+        nonlocal tokens_list
         ret = remove_if_complete(itask)
-        if ret and itask.tokens.task in _tokens_list:
-            _tokens_list.remove(itask.tokens.task)
+        if ret and itask.tokens.task in tokens_list:
+            tokens_list.remove(itask.tokens.task)
         return ret
 
-    schd.pool.remove_if_complete = _remove_if_complete
-
-    # capture workflow shutdown
+    # capture workflow shutdown request
     set_stop = schd._set_stop
-    has_shutdown = False
+    stop_requested = False
 
     def _set_stop(mode=None):
-        nonlocal has_shutdown, stop_mode
+        nonlocal stop_requested, stop_mode
         if mode == stop_mode:
-            has_shutdown = True
+            stop_requested = True
             return set_stop(mode)
         else:
             set_stop(mode)
             raise Exception(f'Workflow bailed with stop mode = {mode}')
 
-    schd._set_stop = _set_stop
-
     # determine the completion condition
-    if _tokens_list:
-        condition = lambda: bool(_tokens_list)
-    else:
-        condition = lambda: bool(not has_shutdown)
-
-    # wait for the condition to be met
-    while condition():
-        # allow the main loop to advance
-        await asyncio.sleep(0)
-        if (time() - start_time) > timeout:
-            raise Exception(
-                f'Timeout waiting for {", ".join(map(str, _tokens_list))}'
-            )
-
-    # restore regular shutdown logic
-    schd._set_stop = set_stop
+    def done():
+        if wait_tokens:
+            return not tokens_list
+        # otherwise wait for the scheduler to shut down
+        if not schd.contact_data:
+            return True
```

Comment:
This fixture was waiting for 60 secs for the workflow to shut down even if there was no contact data. If there is no contact data it must have already shut down due to an error or something (which bypasses the `set_stop` method).